src/pyhf/pdf.py

```diff
@@ -17,90 +17,217 @@
+                raise exceptions.InvalidModel(
+                    'The sample {0:s} has {1:d} bins, but the channel it belongs to ({2:s}) has {3:d} bins.'.format(
+                        sample['name'],
+                        len(sample['data']),
+                        channel['name'],
+                        channel_nbins[channel['name']],
+                    )
+                )
+            for modifier_def in sample['modifiers']:
+                # get the paramset requirements for the given modifier. If
+                # modifier does not exist, we'll have a KeyError
+                try:
+                    paramset_requirements = modifiers.registry[
+                        modifier_def['type']
+                    ].required_parset(len(sample['data']))
+                except KeyError:
+                    log.exception(
+                        'Modifier not implemented yet (processing {0:s}). Available modifiers: {1}'.format(
+                            modifier_def['type'], modifiers.registry.keys()
+                        )
+                    )
+                    raise exceptions.InvalidModifier()
+
+                # check the shareability (e.g. for shapesys for example)
+                is_shared = paramset_requirements['is_shared']
+                if not (is_shared) and modifier_def['name'] in _paramsets_requirements:
+                    raise ValueError(
+                        "Trying to add unshared-paramset but other paramsets exist with the same name."
+                    )
+                if is_shared and not (
+                    _paramsets_requirements.get(
+                        modifier_def['name'], [{'is_shared': True}]
+                    )[0]['is_shared']
+                ):
+                    raise ValueError(
+                        "Trying to add shared-paramset but other paramset of same name is indicated to be unshared."
+                    )
+                _paramsets_requirements.setdefault(modifier_def['name'], []).append(
+                    paramset_requirements
+                )
+    # build up a dictionary of the parameter configurations provided by the user
+    _paramsets_user_configs = {}
+    for parameter in spec.get('parameters', []):
+        if parameter['name'] in _paramsets_user_configs:
+            raise exceptions.InvalidModel(
+                'Multiple parameter configurations for {} were found.'.format(
+                    parameter['name']
+                )
+            )
+        _paramsets_user_configs[parameter.pop('name')] = parameter
+
+    _reqs = reduce_paramsets_requirements(
+        _paramsets_requirements, _paramsets_user_configs
+    )
+
+    _sets = {}
+    for param_name, paramset_requirements in _reqs.items():
+        paramset_type = paramset_requirements.get('paramset_type')
+        paramset = paramset_type(**paramset_requirements)
+        _sets[param_name] = paramset
+    return _sets
```

Comment:
These lines are specific to a given Model. If we care about efficiency and running things once and passing around that object, then we want to factor this out into a separate function called by the Model, with this existing function called by the Workspace, perhaps renaming this as `_paramset_requirements_from_channels`.